### PR TITLE
Update grpc to v1.40.0

### PR DESCRIPTION
**Pull request by yongtang**
This PR updates grpc to v1.40.0

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by yongtang** [comment]
There may still be some hidden tests that may fail.  Though it can be a start. Still need to see if internal tests work or not.

Also not sure about protobuf version in python. In setup.py, 'protobuf >= 3.9.2' has been specified, though it likely requires much later version. Will update if needed.

@satishpasumarthi you can apply the patch to your PR if you prefer. I can close this PR after that.

**Comment by mihaimaruseac** [comment]
I think protobuf and grpc need to be updated in sync. But let's try this way first.

File: tensorflow/workspace2.bzl [hunk 1]
```diff
@@ -560,22 +560,6 @@ def _tf_repositories():
...
-        strip_prefix = "protobuf-3.9.2",
-        system_build_file = "//third_party/systemlibs:protobuf.BUILD",
-        system_link_files = {
-            "//third_party/systemlibs:protobuf.bzl": "protobuf.bzl",
-            "//third_party/systemlibs:protobuf_deps.bzl": "protobuf_deps.bzl",
-        },
-        urls = [
-            "https://storage.googleapis.com/mirror.tensorflow.org/github.com/protocolbuffers/protobuf/archive/v3.9.2.zip",
-            "https://github.com/protocolbuffers/protobuf/archive/v3.9.2.zip",
-        ],
```
Comment (mihaimaruseac) [inline tensorflow/workspace2.bzl:576]: Is this intentional? Do you know why this needs to be removed?

**Comment by yongtang** [comment]
@mihaimaruseac Grpc added a patched version of the protobuffer, I think that might reduce the need to maintain on tensorflow core:

https://github.com/grpc/grpc/blob/master/bazel/grpc_deps.bzl#L212-L223

We can leave protobuffer alone in tensorflow repo. It will override grpc's binding. Though we will need to maintain the patch separately and update protobuffer to a compatible version with grpc.

**Comment by mihaimaruseac** [comment]
Oh, let's go with this one then.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by satishpasumarthi** [comment]
> There may still be some hidden tests that may fail. Though it can be a start. Still need to see if internal tests work or not.
> 
> Also not sure about protobuf version in python. In setup.py, 'protobuf >= 3.9.2' has been specified, though it likely requires much later version. Will update if needed.
> 
> Some windows tests are failing, e.g.,
> 
> ```
> //py_test_dir/tensorflow/python/framework:convert_to_constants_test 
> ```
> 
> I couldn't see if they are related or not (may be related).
> 
> @satishpasumarthi you can apply the patch to your PR #44282 if you prefer. I can close this PR after that.

Are we going with this PR or https://github.com/tensorflow/tensorflow/pull/44282 ?

**Comment by mihaimaruseac** [comment]
I'm ok with either

**Comment by reedwm** [comment]
This is causing the several tests on Windows to fail with the cryptic error: `Windows fatal exception: code 0xc0000374`.

This can be reproduced by running the following in the command prompt in the Tensorflow directory at commit ea3807f1c52984c991e3b832ad462b62c0563a1c:

```bat
configure
set TF2_BEHAVIOR=1
bazel build --announce_rc --config=opt //tensorflow/python/kernel_tests:summary_ops_test && bazel-bin\tensorflow\python\kernel_tests\summary_ops_test.exe SummaryOpsTest.testTrace_withProfiler
```

I plan on rolling this change back soon. Unfortunately, I am unsure what is causing this issue. The output of the test is:

```
[ RUN      ] SummaryOpsTest.testTrace_withProfiler
2021-09-14 19:28:16.764405: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
WARNING:tensorflow:From \\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\ops\summary_ops_v2.py:1305: start (from tensorflow.python.eager.profiler) is deprecated and will be removed after 2020-07-01.
Instructions for updating:
use `tf.profiler.experimental.start` instead.
W0914 19:28:16.815090 17452 deprecation.py:341] From \\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\ops\summary_ops_v2.py:1305: start (from tensorflow.python.eager.profiler) is deprecated and will be removed after 2020-07-01.
Instructions for updating:
use `tf.profiler.experimental.start` instead.
2021-09-14 19:28:16.825582: I tensorflow/core/profiler/lib/profiler_session.cc:110] Profiler session initializing.
2021-09-14 19:28:16.829039: I tensorflow/core/profiler/lib/profiler_session.cc:125] Profiler session started.
WARNING:tensorflow:From \\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\ops\summary_ops_v2.py:1355: stop (from tensorflow.python.eager.profiler) is deprecated and will be removed after 2020-07-01.
Instructions for updating:
use `tf.profiler.experimental.stop` instead.
W0914 19:28:16.940137 17452 deprecation.py:341] From \\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\ops\summary_ops_v2.py:1355: stop (from tensorflow.python.eager.profiler) is deprecated and will be removed after 2020-07-01.
Instructions for updating:
use `tf.profiler.experimental.stop` instead.
2021-09-14 19:28:16.947135: I tensorflow/core/profiler/lib/profiler_session.cc:67] Profiler session collecting data.
2021-09-14 19:28:16.952225: I tensorflow/core/profiler/lib/profiler_session.cc:143] Profiler session tear down.
Windows fatal exception: code 0xc0000374

Current thread 0x0000442c (most recent call first):
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\eager\profiler.py", line 108 in stop
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\util\deprecation.py", line 348 in new_func
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\ops\summary_ops_v2.py", line 1355 in trace_export
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\kernel_tests\summary_ops_test.py", line 1497 in testTrace_withProfiler
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\framework\test_util.py", line 1476 in decorated
  File "C:\Python39\lib\unittest\case.py", line 550 in _callTestMethod
  File "C:\Python39\lib\unittest\case.py", line 593 in run
  File "C:\Python39\lib\unittest\case.py", line 653 in __call__
  File "C:\Python39\lib\unittest\suite.py", line 122 in run
  File "C:\Python39\lib\unittest\suite.py", line 84 in __call__
  File "C:\Python39\lib\unittest\suite.py", line 122 in run
  File "C:\Python39\lib\unittest\suite.py", line 84 in __call__
  File "C:\Python39\lib\unittest\runner.py", line 176 in run
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\absl_py\absl\testing\_pretty_print_reporter.py", line 87 in run
  File "C:\Python39\lib\unittest\main.py", line 271 in runTests
  File "C:\Python39\lib\unittest\main.py", line 101 in __init__
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\absl_py\absl\testing\absltest.py", line 2401 in _run_and_get_tests_result
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\absl_py\absl\testing\absltest.py", line 2430 in run_tests
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\absl_py\absl\testing\absltest.py", line 2106 in _run_in_app
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\absl_py\absl\testing\absltest.py", line 1999 in main
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\platform\googletest.py", line 55 in g_main
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\absl_py\absl\app.py", line 251 in _run_main
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\absl_py\absl\app.py", line 300 in run
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\platform\googletest.py", line 64 in main_wrapper
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\platform\benchmark.py", line 518 in benchmarks_main
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\platform\googletest.py", line 66 in main
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\platform\test.py", line 60 in main
  File "\\?\C:\Users\kbuilder\AppData\Local\Temp\Bazel.runfiles_w37ihgl_\runfiles\org_tensorflow\tensorflow\python\kernel_tests\summary_ops_test.py", line 1625 in <module>
```

**Comment by wchao1115** [comment]
`0xc0000374` is `STATUS_HEAP_CORRUPTION` which usually means apps writing on to a released memory.

**Pull request merged**